common/websocket/index.ts

```diff
@@ -0,0 +1,153 @@
+import { Server } from 'http';
+import { getLogger } from 'log4js';
+import { WebSocketServer, WebSocket } from 'ws';
+import { v4 as createUuid } from 'uuid';
+import { getUserForSession, GraphQLUser } from '../../graphql/authentication';
+
+type UserId = string;
+type ClientId = string;
+
+interface ExtendedWebsocket extends WebSocket {
+    clientId: ClientId;
+    userId: UserId;
+}
+
+type ConnectedClients = Map<UserId, Map<ClientId, ExtendedWebsocket>>;
+
+type Message = {
+    concreteNotificationId: number;
+};
+
+const log = getLogger();
+
+class WebSocketService {
+    private static instance: WebSocketService;
+    private websocketServer: WebSocketServer;
+    private connectedClients: ConnectedClients = new Map();
+
+    private constructor(server: Server) {
+        this.websocketServer = new WebSocketServer({ server, clientTracking: true });
+    }
+
+    public static getInstance(server?: Server) {
+        if (!this.instance && !server) {
+            throw new Error('No instance running and no instantiation possible without a server.');
+        }
+        if (!this.instance) {
+            this.instance = new WebSocketService(server);
+        }
+        return this.instance;
+    }
+
+    getCients(): ConnectedClients {
+        return this.connectedClients;
+    }
+
+    private addClient(userId: string, connection: ExtendedWebsocket) {
+        let userConnections = this.connectedClients.get(userId);
+        if (!userConnections) {
+            userConnections = new Map();
+        }
+        userConnections.set(connection.clientId, connection);
+        this.connectedClients.set(userId, userConnections);
```

Comment:
just a minor thing - this line could go inside the if block above, because otherwise just adding a new connection to an existing user is enough, right?
the naming `this.connectedClients` is misleading as on this level we are dealing with users and not their clients

Reply:
The above if block just creates a new Map storing the user connections if none is present and then adds everything to the connectedUsers :smile:.
I adjusted the naming.